pyxem/generators/indexation_generator.py

```diff
@@ -34,35 +31,48 @@
 
 def correlate_library(image, library, n_largest, mask, keys=[]):
     """Correlates all simulated diffraction templates in a DiffractionLibrary
-    with a particular experimental diffraction pattern (image) stored as a
-    numpy array. See the correlate method of IndexationGenerator for details.
+    with a particular experimental diffraction pattern (image).
+
+    Parameters
+    ----------
+    image : np.array()
+        The experimental diffraction pattern of interest.
+    library : DiffractionLibrary
+        The library of diffraction simulations to be correlated with the
+        experimental data.
+    n_largest : int
+        The number of well correlated simulations to be retained.
+    mask : bool array
+        A mask for navigation axes 1 indicates positions to be indexed.
+
+    Returns
+    -------
+    top_matches : (<num phases>*n_largest, 5), np.array()
+        A numpy array containing the top n correlated simulations for the
+        experimental pattern of interest.
+
+    See also
+    --------
+    pyxem.utils.correlate and the correlate method of IndexationGenerator.
     """
-    i = 0
-    out_arr = np.zeros((n_largest * len(library), 5))
+    top_matches = np.zeros((len(library), n_largest, 5))
     if mask == 1:
-        for key in library.keys():
-            correlations = dict()
-            for orientation, diffraction_pattern in library[key].items():
-                # diffraction_pattern here is in fact a library of
-                # diffraction_pattern_properties
+        for phase_index, key in enumerate(library.keys()):
+            correlations = np.empty((len(library[key]), 4))
+            for i, (orientation, diffraction_pattern) in enumerate(library[key].items()):
```

Comment:
For clarity I would prefer 
```
i=0
for orientation,diffraction_pattern in library[keys].items():
    ###
    i+=1
```
but i'll leave it at your discretion

Reply:
From python 3.6 (CPython implementation), dict enumeration is guaranteed to be in insertion order. From 3.7 it is a general language guarantee. Also: We are partitioning after the loop, so this would only be a problem for ties at the partition point?